update/sync/coordinator.go

```diff
@@ -0,0 +1,481 @@
+	ss.transactions.storage[block.RewardsBlock] = args.Storages.GetStorer(dataRetriever.RewardTransactionUnit)
+
+	for _, pool := range ss.transactions.txPools {
+		pool.RegisterHandler(ss.receivedTransaction)
+	}
+
+	for _, store := range ss.transactions.storage {
+		if check.IfNil(store) {
+			return nil, epochStart.ErrNilStorage
+		}
+	}
+
+	ss.transactions.chReceivedAll = make(chan bool)
+	ss.miniBlocks.chReceivedAll = make(chan bool)
+
+	ss.miniBlocks.pool.RegisterHandler(ss.receivedMiniBlock)
+
+	return ss, nil
+}
+
+// SyncAllState gets an epoch number and will sync the complete data for that epoch start metablock
+func (ss *syncState) SyncAllState(epoch uint32) error {
+	if epoch == ss.syncingEpoch {
+		return nil
+	}
+	ss.syncingEpoch = epoch
+
+	meta, err := ss.headers.SyncEpochStartMetaHeader(epoch, time.Minute)
+	if err != nil {
+		return err
+	}
+
+	ss.syncingEpoch = meta.GetEpoch()
+
+	wg := sync.WaitGroup{}
+	wg.Add(len(meta.EpochStart.LastFinalizedHeaders) + 1)
+
+	var errFound error
+	mutErr := sync.Mutex{}
+
+	go func() {
+		errMeta := ss.syncMeta(meta, &wg)
+		if errMeta != nil {
+			mutErr.Lock()
+			errFound = errMeta
+			mutErr.Unlock()
+		}
+	}()
+
+	for _, shData := range meta.EpochStart.LastFinalizedHeaders {
+		go func(shardData block.EpochStartShardData) {
+			err := ss.syncShard(shardData, &wg)
+			if err != nil {
+				mutErr.Lock()
+				errFound = err
+				mutErr.Unlock()
+			}
+		}(shData)
+	}
+
+	wg.Wait()
+
+	if errFound != nil {
+		return errFound
+	}
+
+	return nil
+}
+
+func (ss *syncState) syncShard(shardData block.EpochStartShardData, wg *sync.WaitGroup) error {
+	defer wg.Done()
+	err := ss.syncTrieOfType(factory.UserAccount, shardData.ShardId, shardData.RootHash)
+	if err != nil {
+		return err
+	}
+
+	_ = process.EmptyChannel(ss.transactions.chReceivedAll)
+	_ = process.EmptyChannel(ss.miniBlocks.chReceivedAll)
+
+	ss.miniBlocks.mutPendingMb.Lock()
+	ss.transactions.mutPendingTx.Lock()
+
+	requestedMBs := 0
+	requestedTxs := 0
+	for _, mbHeader := range shardData.PendingMiniBlockHeaders {
+		ss.miniBlocks.mapHashes[string(mbHeader.Hash)] = struct{}{}
+		miniBlock, ok := ss.getMiniBlockFromPoolOrStorage(mbHeader.Hash)
+		if ok {
+			ss.miniBlocks.mapMiniBlocks[string(mbHeader.Hash)] = miniBlock
+			for _, txHash := range miniBlock.TxHashes {
+				ss.transactions.mapHashes[string(txHash)] = miniBlock
+			}
+			requestedTxs += ss.requestTransactionsForMB(miniBlock)
+			continue
+		}
+
+		requestedMBs++
+		ss.requestHandler.RequestMiniBlock(mbHeader.SenderShardID, mbHeader.Hash)
+	}
+
+	ss.transactions.mutPendingTx.Unlock()
+	ss.miniBlocks.mutPendingMb.Unlock()
+
+	if requestedMBs > 0 {
+		err := WaitFor(ss.miniBlocks.chReceivedAll, time.Hour)
+		log.Warn("could not finish syncing", "error", err)
+		if err != nil {
+			return err
+		}
+	}
+	if requestedTxs > 0 {
+		err := WaitFor(ss.transactions.chReceivedAll, time.Hour)
```

Comment:
Is this an established good time frame to wait? Should it be configurable?

Reply:
will think about how to make it configurable. maybe changing it to haveTime and give a global waiting time for all.